File: backend/app/schemas/loans.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Loan Requests
# ---------------------------------------------------------------------------

class LoanRequestCreate(BaseModel):
    message: Optional[str] = None


class LoanRequestApprove(BaseModel):
    agreed_deposit: Decimal = Decimal("0")
    meet_location: str
    due_days: int = 14  # how many days the borrower gets


class LoanRequestReject(BaseModel):
    reason: Optional[str] = None


class UserSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    avatar_url: Optional[str] = None
    phone_verified: bool


class BookSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    cover_url: Optional[str] = None
    authors: Optional[list[str]] = None


class LoanRequestOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_book_id: uuid.UUID
    lender_id: uuid.UUID
    borrower_id: uuid.UUID
    message: Optional[str] = None
    status: str
    agreed_deposit: Optional[Decimal] = None
    meet_location: Optional[str] = None
    agreed_at: Optional[datetime] = None
    rejected_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    lender: UserSummary
    borrower: UserSummary


# ---------------------------------------------------------------------------
# Loans
# ---------------------------------------------------------------------------

class LoanOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    loan_request_id: uuid.UUID
    user_book_id: uuid.UUID
    lender_id: uuid.UUID
    borrower_id: uuid.UUID
    lent_at: date
    due_at: Optional[date] = None
    returned_at: Optional[date] = None
    status: str
    lender_note: Optional[str] = None
    created_at: datetime
    lender: UserSummary
    borrower: UserSummary

File: frontend/src/api/loans.ts
import apiClient from "./client";

export interface UserSummary {
  id: string;
  name: string;
  avatar_url: string | null;
  phone_verified: boolean;
}

export interface LoanRequest {
  id: string;
  user_book_id: string;
  lender_id: string;
  borrower_id: string;
  message: string | null;
  status: string;
  agreed_deposit: string | null;
  meet_location: string | null;
  agreed_at: string | null;
  rejected_reason: string | null;
  created_at: string;
  updated_at: string;
  lender: UserSummary;
  borrower: UserSummary;
}

export interface Loan {
  id: string;
  loan_request_id: string;
  user_book_id: string;
  lender_id: string;
  borrower_id: string;
  lent_at: string;
  due_at: string | null;
  returned_at: string | null;
  status: string;
  lender_note: string | null;
  created_at: string;
  lender: UserSummary;
  borrower: UserSummary;
}

export interface Notification {
  id: string;
  type: string;
  title: string;
  body: string | null;
  is_read: boolean;
  content_type: string | null;
  content_id: string | null;
  created_at: string;
}

export const loansApi = {
  requestLoan: (bookId: string, message?: string) =>
    apiClient.post<LoanRequest>(`/books/${bookId}/request-loan`, { message }),

  listIncoming: (status?: string) =>
    apiClient.get<LoanRequest[]>("/loan-requests", { params: status ? { status } : {} }),

  listSent: () => apiClient.get<LoanRequest[]>("/loan-requests/sent"),

  approve: (id: string, data: { agreed_deposit: number; meet_location: string; due_days?: number }) =>
    apiClient.put<LoanRequest>(`/loan-requests/${id}/approve`, data),

  reject: (id: string, reason?: string) =>
    apiClient.put<LoanRequest>(`/loan-requests/${id}/reject`, { reason }),

  cancel: (id: string) => apiClient.delete(`/loan-requests/${id}`),

  listLoans: (status?: string) =>
    apiClient.get<Loan[]>("/loans", { params: status ? { status } : {} }),

  confirm: (loanRequestId: string, due_days = 14) =>
    apiClient.put<Loan>(`/loans/${loanRequestId}/confirm`, null, { params: { due_days } }),

  markReturned: (loanId: string, note?: string) =>
    apiClient.put<Loan>(`/loans/${loanId}/return`, null, { params: note ? { note } : {} }),
};

export const notificationsApi = {
  list: () => apiClient.get<Notification[]>("/notifications"),
  markRead: (id: string) => apiClient.put(`/notifications/${id}/read`),
  markAllRead: () => apiClient.put("/notifications/read-all"),
};
